policies/recipes/steps.go

```diff
@@ -90,10 +97,274 @@ func parsePermissions(s string) (os.FileMode, error) {
+		}
+		if zipIsDir(f.Name) && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	// Create files.
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, normalizeSlashes(f.Name)))
+		if err != nil {
+			return err
+		}
+		if zipIsDir(f.Name) {
+			mode := f.Mode()
+			if mode == 0 {
+				mode = 0755
+			}
+			if err = os.MkdirAll(filen, mode); err != nil {
+				return err
+			}
+			continue
+		}
+		filedir := filepath.Dir(filen)
+		if err = os.MkdirAll(filedir, 0755); err != nil {
+			return err
+		}
+		fmt.Printf("os.Create %s\n", filen)
+		reader, err := f.Open()
+		if err != nil {
+			return err
+		}
+
+		mode := f.Mode()
+		if mode == 0 {
+			mode = 0755
+		}
+
+		dst, err := os.OpenFile(filen, os.O_RDWR|os.O_CREATE, mode)
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(dst, reader)
+		dst.Close()
+		reader.Close()
+		if err != nil {
+			return err
+		}
+		err = os.Chtimes(filen, time.Now(), f.ModTime())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func decompress(reader io.Reader, archiveType osconfigpb.SoftwareRecipe_Step_ExtractArchive_ArchiveType) (io.Reader, error) {
+	switch archiveType {
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_GZIP:
+		// *gzip.Reader is a io.ReadCloser but it isn't necesary to call Close() on it.
+		return gzip.NewReader(reader)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_BZIP:
+		return bzip2.NewReader(reader), nil
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_LZMA:
+		return lzma.NewReader2(reader)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_XZ:
+		return xz.NewReader(reader)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR:
+		return reader, nil
+	default:
+		return nil, fmt.Errorf("Unrecognized archive type %q when trying to decompress tar", archiveType)
+	}
+}
+
+func checkForConflicts(tr *tar.Reader, dst string) error {
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if header.Typeflag == tar.TypeDir && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
 	return nil
 }
 
+func createFiles(tr *tar.Reader, dst string) error {
+	for {
+		var err error
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		filedir := filepath.Dir(filen)
+		err = os.MkdirAll(filedir, 0755)
+		if err != nil {
+			return err
+		}
+		switch header.Typeflag {
+		case tar.TypeDir:
```

Comment:
The ordering of directories isn't set either. My current plan is to just create directories as needed (I set the permissions on them to 700 initially to prevent other people from reading them) and then change the permissions and owner of them to be correct when I get to the correct record in the tar.

Reply:
ok, let's see if this works out for us